src/app/modules/payments/stripe.utils.ts

```diff
@@ -54,25 +93,42 @@ const isStripeMetadata = (
  * it (i.e. payment intents and charges).
  * @param {Stripe.Metadata} metadata the metadata object which is expected to have a payment id
  * @returns ok(paymentId) the extracted paymentId
+ * @returns err(StripeMetadataInvalidError) if the metadata has an invalid shape
  * @returns err(StripeMetadataValidPaymentIdNotFoundError) if the payment id was not found or is an invalid BSON object id
+ * @returns err(StripeMetadataIncorrectEnvError) if the app is incorrect
  */
 export const getMetadataPaymentId = (
   metadata: Stripe.Metadata,
-): Result<string, StripeMetadataValidPaymentIdNotFoundError> => {
+): Result<
+  string,
+  | StripeMetadataInvalidError
+  | StripeMetadataValidPaymentIdNotFoundError
+  | StripeMetadataIncorrectEnvError
+> => {
   const logMeta = {
     action: 'getMetadataPaymentId',
     metadata,
   }
-  if (
-    !isStripeMetadata(metadata) ||
-    !mongoose.Types.ObjectId.isValid(metadata.paymentId)
-  ) {
+  if (!isStripeMetadata(metadata)) {
     logger.warn({
-      message: 'Got metadata with invalid paymentId',
+      message: 'Got invalid Stripe metadata',
+      meta: { ...logMeta, metadata },
+    })
+    return err(new StripeMetadataInvalidError())
+  }
+  if (!mongoose.Types.ObjectId.isValid(metadata.paymentId)) {
+    logger.warn({
+      message: 'Got Stripe metadata with invalid paymentId',
       meta: { ...logMeta, metadata },
     })
     return err(new StripeMetadataValidPaymentIdNotFoundError())
   }
+  // Explicit check for metadata.env to ensure that non-existent legacy metadata
+  // passes through on production.
```

Comment:
To clarify: do you mean to ensure that an error (ie `StripeMetadataIncorrectEnvError`) is thrown for legacy metadata?

Reply:
maybe this was a bad description. In the legacy version, `env` does not exist. probably a better comment would have been `legacy metadata with non-existent env value` 😅

Reply:
clarified in [d33087a](https://github.com/opengovsg/FormSG/pull/6234/commits/d33087a09a2865a5c465c4bdfec2bdb9fc999061)